felix/daemon/daemon.go

```diff
@@ -368,13 +368,24 @@ configRetry:
 		simulateDataRace()
 	}
 
-	// We may need to temporarily disable encrypted traffic to this node in order to connect to Typha
-	if configParams.WireguardEnabled {
-		err := bootstrapWireguard(configParams, v3Client)
-		if err != nil {
-			time.Sleep(2 * time.Second) // avoid a tight restart loop
-			log.WithError(err).Fatal("Couldn't bootstrap WireGuard host connectivity")
-		}
+	// Perform wireguard bootstrap processing. This may remove wireguard configuration if wireguard is disabled or
+	// if the configuration is obviously broken. This also filters the typha addresses based on whether routing is
+	// obviously broken to the typha node (due to wireguard routing asymmetry). If we end up filtering out all of the
+	// typha addresses then we will need to remove our wireguard configuration to proceed.
+	typhaAddresses, err := bootstrapWireguardAndFilterTyphaAddresses(configParams, v3Client, typhaAddresses)
+	if err != nil {
+		time.Sleep(2 * time.Second) // avoid a tight restart loop
+		log.WithError(err).Fatal("Couldn't bootstrap WireGuard host connectivity")
+	}
+
+	felixConfiguredToUseTypha := configParams.TyphaK8sServiceName != ""
```

Comment:
I _think_ this code is unreachable since bootstrapWireguardAndFilterTyphaAddresses will return an error if if filters away _all_ the typha addresses.  That said, I think relying on that detail of bootstrapWireguardAndFilterTyphaAddresses feels a bit brittle.

The check you've got is slightly incomplete since felix can be configured to talk to Typha with TyphaK8sServiceName or TyphaAddr, I think.  If we keep this, I suggest initialising `felixConfiguredToUseTypha` above before the filtering happens by doing `felixConfiguredToUseTypha = len(typhaAddresses) > 0`

Reply:
Yeah I don't think this is necessary although reasons are a little different.

The bootstrap code should never return 0 addresses unless it was given 0 addresses in the first place. If it filters out all addresses (because of wireguard) it will remove wireguard and then return the full set of typha addresses.

Reply:
I'll update the comment on the Bootstrap function